lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -162,13 +162,16 @@ def find_matches(self, requirements):
                 for version, _none_src_server in coll_versions
             ]
 
-        preinstalled_candidates = {
-            Candidate(req.fqcn, req.ver, req.src, req.type)
-            for req in self._preferred_requirements
-            if req.fqcn == fqcn
-        }
+        all_preinstalled_candidates = set()
+        preinstalled_candidates = set()
+        for req in self._preferred_requirements:
+            if req.fqcn == fqcn:
+                candidate = Candidate(req.fqcn, req.ver, req.src, req.type)
+                all_preinstalled_candidates.add(candidate)
+                if all (self.is_satisfied_by(r, candidate) for r in requirements):
+                    preinstalled_candidates.add(candidate)
```

Comment:
There wasn't a unit test that caught this. I'm going to add an integration test against devel in ansible-galaxy-collection.

A simple reproducer for the bug:
Start with amazon.aws not installed
Install version 1.0.0:
`$ ansible-galaxy collection install amazon.aws:1.0.0`
Install a different version:
`$ ansible-galaxy collection install amazon.aws:1.3.0`

Results in:
```
Traceback (most recent call last):
  File "/home/fedora/ansible/bin/ansible-galaxy", line 135, in <module>
    exit_code = cli.run()
  File "/home/fedora/ansible/lib/ansible/cli/galaxy.py", line 536, in run
    context.CLIARGS['func']()
  File "/home/fedora/ansible/lib/ansible/cli/galaxy.py", line 68, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/home/fedora/ansible/lib/ansible/cli/galaxy.py", line 1127, in execute_install
    artifacts_manager=artifacts_manager,
  File "/home/fedora/ansible/lib/ansible/cli/galaxy.py", line 1154, in _execute_install_collection
    artifacts_manager=artifacts_manager,
  File "/home/fedora/ansible/lib/ansible/galaxy/collection/__init__.py", line 472, in install_collections
    allow_pre_release=allow_pre_release,
  File "/home/fedora/ansible/lib/ansible/galaxy/collection/__init__.py", line 1221, in _resolve_depenency_map
    max_rounds=2000000,  # NOTE: same constant pip uses
  File "/home/fedora/.local/lib/python3.7/site-packages/resolvelib/resolvers.py", line 414, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/fedora/.local/lib/python3.7/site-packages/resolvelib/resolvers.py", line 312, in resolve
    failure_causes = self._attempt_to_pin_criterion(name, criterion)
  File "/home/fedora/.local/lib/python3.7/site-packages/resolvelib/resolvers.py", line 226, in _attempt_to_pin_criterion
    raise InconsistentCandidate(candidate, criterion)
resolvelib.resolvers.InconsistentCandidate: Provided candidate <amazon.aws:1.0.0 of type 'dir' from b'/home/fedora/ansible/test_collections/ansible_collections/amazon/aws'> does not satisfy <amazon.aws:1.3.0 of type 'galaxy' from Galaxy>
```

Reply:
Actually, I guess the test needs to be added to this PR, since on devel it requires --force which bypasses this bug.

Reply:
Interesting. Sounds like we could attempt to do an early exit for pre-installed collections up the stack to preserve the behavior and get back to it in a separate PR.

Reply:
Works for me :+1: